samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,751 @@
+	 NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+/* Event flags */
+#define NET_IPV4_EVENT_READY BIT(1)
+#define NET_IPV6_EVENT_READY BIT(2)
+
+/* OTA update state machine states */
+enum ota_state {
+	OTA_STATE_SCAN,           /**< Scan for WiFi networks */
+	OTA_STATE_SERVER_CONNECT, /**< Connect to OTA server */
+	OTA_STATE_DOWNLOAD        /**< Download and process firmware image */
+};
+
+struct http_parse_data {
+	char *schema;
+	char *host;
+	char *path;
+	char *port;
+};
+
+struct app_ctx {
+	uint16_t scan_result_count;
+	enum wifi_security_type security_type;
+	struct k_event events;
+	struct net_mgmt_event_callback ipv4_mgmt_cb;
+	struct net_mgmt_event_callback ipv6_mgmt_cb;
+	struct net_mgmt_event_callback wlan_mgmt_cb;
+	struct net_mgmt_event_callback dns_mgmt_cb;
+	volatile enum ota_state state;
+	volatile bool ipv4_addr_config_done;
+	volatile bool ipv6_addr_config_done;
+	volatile bool disonnected;
+	int sock;
+	struct http_parse_data http_parse_info;
+	uint32_t ota_image_size;
+	int ota_range_start_byte;
+	int ota_range_end_byte;
+	int http_status_code;
+
+	/* Flash buffer for firmware chunks */
+	uint8_t flash_buffer[1024];
+	uint32_t flash_buffer_len;
+
+	/* Separate HTTP receive buffer to avoid conflicts */
+	uint8_t http_recv_buffer[1024];
+
+	uint8_t retry_count;
+};
+
+static char *ota_mac_to_string(const uint8_t *mac, uint8_t mac_len, char *buf, size_t buf_len)
+{
+	if (!mac || mac_len != 6 || !buf || buf_len < 18) {
+		return NULL;
+	}
+	snprintf(buf, buf_len, "%02x:%02x:%02x:%02x:%02x:%02x", mac[0], mac[1], mac[2], mac[3],
+		 mac[4], mac[5]);
+	return buf;
+}
+
+static char *ota_get_url_field(const char *url, struct http_parser_url parser, int field)
+{
+	int len = parser.field_data[field].len;
+	const char *ptr = url + parser.field_data[field].off;
+	char *out;
+
+	out = strndup(ptr, len);
+	return out;
+}
+
+/*
+ * @brief Parses URL into host, path and port componentss
+ * @param http_parse_info Pointer to structure to store the parsed URL components
+ * @return 0 on success, negative error code on failure
+ */
+static int ota_parse_url(struct http_parse_data *http_parse_info)
+{
+	char *full_url = CONFIG_OTA_UPDATE_URL;
+	struct http_parser_url parser;
+	int ret = 0;
+
+	http_parser_url_init(&parser);
+	ret = http_parser_parse_url(full_url, strlen(full_url), 0, &parser);
+	__ASSERT(ret == 0, "URL parsing failed");
+
+	/* Get the schema, host, port and path info from the parsed URL */
+	__ASSERT(parser.field_set & BIT(UF_SCHEMA), "Schema field missing in URL");
+	__ASSERT(parser.field_set & BIT(UF_HOST), "Host field missing in URL");
+	__ASSERT(parser.field_set & BIT(UF_PATH), "Path field missing in URL");
+
+	http_parse_info->schema = ota_get_url_field(full_url, parser, UF_SCHEMA);
+	http_parse_info->host = ota_get_url_field(full_url, parser, UF_HOST);
+	http_parse_info->path = ota_get_url_field(full_url, parser, UF_PATH);
+	http_parse_info->port = ota_get_url_field(full_url, parser, UF_PORT);
```

Comment:
This line is useless since `http_parse_info->port` is set below. Worse, it generate a memory leak because of `strdup()`.

Reply:
This is done in the `else` branch below.

Reply:
BTW, I am not sure we need to call `strdup()`. Maybe we could add `'\0'` directly in the URL string. We just need to make sure the URL won't be in `const`:

        char url[] = CONFIG_OTA_URL;
        [...]
        int len = parser.field_data[field].len;
        char *ptr = url + parser.field_data[field].off;

        __ASSERT(parser.field_set & BIT(field), "invalid URL");
        ptr[len] = '\0';

Reply:
@asbarnwa, why do we need the memory allocation rather than constructing the url components in-place?